lite/src/sequential.rs

```diff
@@ -0,0 +1,102 @@
+{
+    /// Returns the time after which the current state expires
+    /// and the verifier must be reset subjectively.
+    fn expires(&self) -> Time {
+        self.state.header.bft_time() + self.trusting_period
+    }
+
+    /// Verify takes a header, a commit for the header,
+    /// and the next validator set referenced by the header.
+    /// Without knowing this next validator set, we can't
+    /// really verify the next header, so we make verifying
+    /// the current header conditional on receiving that validator set.
+    /// We could break this up into two steps, one where we verify the header,
+    /// and one where we receive and validate the next validators,
+    /// but for now we combine them. Returns an error if verification fails.
+    fn verify<C>(&self, now: Time, header: H, commit: C, next_validators: V) -> Result<(), Error>
+    where
+        C: Commit,
+    {
+        // ensure the state is not expired.
+        if self.expires() < now {
+            return Err(Error::Expired);
+        }
+
+        // ensure the height is strictly sequential.
+        if header.height() != self.state.header.height() + 1 {
+            return Err(Error::NonSequentialHeight);
+        }
+
+        // ensure the validators in the header matches what we expect from our state.
+        if header.validators_hash() != self.state.next_validators.hash() {
+            return Err(Error::InvalidValidators);
+        }
+
+        // ensure the next validators in the header matches what was supplied.
+        if header.next_validators_hash() != next_validators.hash() {
+            return Err(Error::InvalidNextValidators);
+        }
+
+        // ensure the commit matches the header.
+        if header.hash() != commit.header_hash() {
+            return Err(Error::InvalidCommitValue);
+        }
+
+        // ensure that +2/3 validators signed correctly
+        self.verify_commit(commit)
+    }
+
+    /// Check that +2/3 of the correct and trusted validator set signed this commit.
+    fn verify_commit<C>(&self, commit: C) -> Result<(), Error>
+    where
+        C: Commit,
+    {
+        let total_power = self.state.next_validators.total_power();
+        let mut signed_power: u64 = 0;
+
+        // NOTE that the vals and commit have a 1-to-1 correspondance.
+        // This means we don't need the validator IDs or to do any lookup,
+        // we can just zip the iterators.
+        let vals_iter = self.state.next_validators.into_vec().into_iter();
+        let commit_iter = commit.into_vec().into_iter();
+        for (val, vote_opt) in vals_iter.zip(commit_iter) {
+            // skip absent and nil votes
+            // NOTE: do we want to check the validity of votes
+            // for nil ?
+            let vote = match vote_opt {
+                Some(v) => v,
+                None => continue,
+            };
+
+            // check vote is valid from validator
+            if !val.verify(vote.sign_bytes(), vote.signature()) {
```

Comment:
Where do we verify that signed bytes corresponds to the right BlockID?

Reply:
We check that the `commit.header_hash()` matches the hash of the header. This header is a component of the block id. It's the responsibility of the implementor of vote to ensure the sign bytes include the correct block id. I thought I could get away with not even surfacing the block id here, since the light client really only cares about the header hash, the block id is kind of an implementation detail, but i may have been a bit overzealous ...

Reply:
I am worried with the case that vote looks like correct one but it is not for the right blockID. Don't have precise scenario yet.

Reply:
Like it's for the right header but the wrong block parts ? I am worried about it too, but in this design it's supposed to be taken care of by the code that implements the traits for real data structures - ie. it's that code's responsibility to only include votes that have the right block id, so by the time this code runs, we can assume that all votes are for the right block id.

Reply:
Like it's for the right header but the wrong block parts ? I am worried about it too, but in this design it's supposed to be taken care of by the code that implements the traits for real data structures - ie. it's that code's responsibility to only include votes that have the right block id, so by the time this code runs, we can assume that all votes are for the right block id.

Reply:
Added a note about this in the Vote comments